### PR TITLE
[dtensor][2/N] more tensor ops to use strategy propagation

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101203

As titled, this PR adapts a few more tensor ops to use strategy based
sharding prop

### CODE REVIEW

File: torch/distributed/_tensor/ops/tensor_ops.py [hunk 1]
```diff
@@ -203,85 +198,10 @@ def replicate_tensor_dim(
...
-    if start == 0 and end == input_spec.shape[dim] and step == 1:
-        return OutputSharding(output_spec=input_spec)
-
-    return _prop_all_but_dim(op_schema, dim=dim)
-
-
-@register_prop_rule(aten.slice_scatter.default)
-def prop_slice_scatter(op_schema: OpSchema) -> OutputSharding:
+@register_op_strategy(aten.slice_scatter.default)
+def gen_slice_scatter_strategy(
```
Comment (XilunWu) [inline torch/distributed/_tensor/ops/tensor_ops.py:202]: question: can we merge the common logic with `gen_slice_strategy`?

File: torch/distributed/_tensor/ops/tensor_ops.py [hunk 2]
```diff
@@ -290,60 +210,47 @@ def prop_slice_scatter(op_schema: OpSchema) -> OutputSharding:
...
-            output_spec=DTensorSpec(
-                mesh=input.mesh,
-                placements=input.placements,
+    if not slice_scatter_strategy.strategies:
+        # if all strategies are filtered out, replicating all specs on slice_scatter dim
+        # of the input strategy, and use that as the op strategy
+        for arg_strategy in input_strategy.strategies:
+            arg_spec = arg_strategy.output_spec
+            replicate_spec = DTensorSpec(
+                mesh, replicate_tensor_dim(arg_spec.placements, dim=slice_dim)
```
Comment (XilunWu) [inline torch/distributed/_tensor/ops/tensor_ops.py:239]: the difference between `replicate_tensor_dim` and `unshard_tensor_dim` is, `unshard_tensor_dim` only replicate when the dim is `Shard` while `replicate_tensor_dim` replicate when the dim is either `Shard` or `_Partial`. What is the reason doing this?

File: torch/distributed/_tensor/ops/utils.py [hunk 3]
```diff
@@ -69,3 +70,33 @@ def normalize_dims(dims: Union[int, Sequence[int]], ndim: int) -> Sequence[int]:
...
+    # number of shards in each tensor dimension
+    shards_map = [1] * len(shape)
+    for i, placement in enumerate(spec.placements):
+        if placement.is_shard():
+            shard_dim = cast(Shard, placement).dim
+            shards_map[shard_dim] *= spec.mesh.size(i)
+
+    for i, dim_size in enumerate(shape):
+        # TODO: maybe we should determine is_sharable base on the
+        #      whether it's evenly sharded or not
```
Comment (XilunWu) [inline torch/distributed/_tensor/ops/utils.py:88]: Is the indentation correct here? Can polish the `TODO` information:
"is_sharable" -> "is_shardable"
"base" -> "based"
"on the whether" -> "on whether"

**Review by XilunWu — APPROVED** [review]
lgtm! some nits and questions.

**Review by fduwjj — COMMENTED** [review]
Do we have a design or RFC for the op strategy propagation?

File: torch/distributed/_tensor/ops/tensor_ops.py [hunk 4]
```diff
@@ -290,60 +210,47 @@ def prop_slice_scatter(op_schema: OpSchema) -> OutputSharding:
...
-        )
+            slice_scatter_strategy.strategies.append(
+                PlacementStrategy(output_spec=replicate_spec)
+            )
+    return slice_scatter_strategy
+
+
+@register_op_strategy(aten._local_scalar_dense.default)
+def replica_only_strategy(
+    node: Node, mesh: DeviceMesh, node_to_strategy: Dict[Node, StrategyType]
```
Comment (fduwjj) [inline torch/distributed/_tensor/ops/tensor_ops.py:249]: ditto.

File: torch/distributed/_tensor/ops/utils.py [hunk 5]
```diff
@@ -69,3 +70,33 @@ def normalize_dims(dims: Union[int, Sequence[int]], ndim: int) -> Sequence[int]:
...
+        if dim_size < shards_map[i]:
+            return False
```
Comment (fduwjj) [inline torch/distributed/_tensor/ops/utils.py:90]: Is this to avoid cases when one rank has no local tensor?

**Review by fduwjj — COMMENTED** [review]
Just another n00b question, can eager mode also leverage this work or we definitely need to first have a fx trace graph first?

**Review by fduwjj — COMMENTED** [review]
Just another n00b question, can eager mode also leverage this work or we definitely need to first have a fx trace graph first?

File: torch/distributed/_tensor/ops/tensor_ops.py (hunk 2 above)
Comment (wanchaol) [reply torch/distributed/_tensor/ops/tensor_ops.py:240]: `replicate_tensor_dim` means if tensor dim is sharded we replicate it, also not allowing partial, `unshard` tensor dim only checks if it's sharded but not check partial, this is to deal with cases are not accepting partial, i.e. `slice_scatter` not accepting partial

**Comment by wanchaol** [comment]
> Just another n00b question, can eager mode also leverage this work or we definitely need to first have a fx trace graph first?

The fx graph is on the op level, used to propagate the tensor metadata and sharding, both eager and graph could leverge this

**Pull request closed without merging**